File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/model/LocalModeHelper.java
package de.dimedis.mobileentry.model;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Optional;
import java.util.jar.Attributes;
import java.util.jar.JarFile;
import java.util.jar.Manifest;

import dalvik.system.DexClassLoader;
import de.dimedis.lmlib.LocalModeImpl;
import de.dimedis.lmlib.myinterfaces.Access;
import de.dimedis.lmlib.myinterfaces.LocalMode;
import de.dimedis.mobileentry.ConfigPref;
import de.dimedis.mobileentry.MobileEntryApplication;
import de.dimedis.mobileentry.backend.response.Border;
import de.dimedis.mobileentry.util.AppContext;
import de.dimedis.mobileentry.util.CommonUtils;
import de.dimedis.mobileentry.util.ConfigPrefHelper;
import de.dimedis.mobileentry.util.Logger;
import de.dimedis.mobileentry.util.PrefUtils;

public class LocalModeHelper {
    private static final String TAG = LocalModeHelper.class.getSimpleName();
    private static final String LOCAL_MODE_CLASS_NAME = "de.dimedis.lmlib.impl.LocalModeImpl";
    private static final String LM_LIB_NAME = "lmlib.jar";
    private static LocalMode sLocalMode;

    private LocalModeHelper() {
    }

    public static Access performEntry(@NonNull String ticketCode) {
        return performEntry(new PerfData().initFromPref(), ticketCode);
    }

    public static Access performEntry(@NonNull PerfData pref, @NonNull String ticketCode) {

        LocalMode lm = getLocalMode();
        if (lm != null) {
            try {
                Access access = lm.performEntry(ticketCode, pref.getDeviceSuid(), pref.getUserSession(), pref.getUserSuid(), pref.getUserName(), pref.getFair(), pref.getBorder(), pref.getLang());
                Log.w(TAG, "access: " + access);
                return access;
            } catch (Exception e) {
                Logger.w(TAG, "Failed to perform entry", e);
            }
        }
        return null;
    }

    public static Access performCheckout(@NonNull String ticketCode) {
        return performCheckout(new PerfData().initFromPref(), ticketCode);
    }

    public static Access performCheckout(@NonNull PerfData pref, @NonNull String ticketCode) {
        LocalMode lm = getLocalMode();
        if (lm != null) {

            try {
                return lm.performCheckout(ticketCode, pref.getDeviceSuid(), pref.getUserSession(), pref.getUserSuid(), pref.getUserName(), pref.getFair(), pref.getBorder(), pref.getLang());
            } catch (Exception e) {
                Logger.w(TAG, "Failed to perform checkout", e);
            }
        }
        return null;
    }

    public static Access recordEntry(@NonNull String ticketCode) {
        LocalMode lm = getLocalMode();
        if (lm != null) {
            Border border = ConfigPrefHelper.getUsersBorder();
            if (border != null) {
                ConfigPref pref = MobileEntryApplication.getConfigPreferences();
                try {
                    return lm.recordEntry(ticketCode, pref.deviceSuid(), pref.userSession(), pref.userSuid(), pref.userName(), border.getFair(), border.getBorder(), pref.currentLanguage());
                } catch (Exception e) {
                    Logger.w(TAG, "Failed to record entry", e);
                }
            }
        }
        return null;
    }

    public static LocalMode getLocalMode() {
        if (sLocalMode == null) {
            sLocalMode = load();
        }
        return sLocalMode;
    }

    public static String getLocalLibVersionFromManifestFile(File fullPathToJar) {
        try {
            JarFile jar = new JarFile(fullPathToJar);
            final Manifest manifest = jar.getManifest();
            final Attributes mattr = manifest.getMainAttributes();
            String version = mattr.getValue("Implementation-Version");
            System.out.println("\\nVersion" + version + ":\\n");
            if (version == null) {
                throw new IOException("unable to read jar version from manifest");
            }
            return version;
        } catch (FileNotFoundException fex) {
            Log.e(TAG, "unable to find lib file: " + fullPathToJar.getAbsolutePath());
        } catch (IOException ex) {
            ex.printStackTrace();
        }
        return null;
    }

    static void initDefConfig() {
        InputStream is = null;
        try {
            is = AppContext.get().getAssets().open("offline-test-config.json");
            String conf = CommonUtils.inputStreamToString(is);
            ConfigPrefHelper.getPref().setOfflineConfig(conf);
            Log.w(TAG, "initDefConfig conf size: " + (conf != null ? conf.length() : "null") + " data:" + conf);
        } catch (Throwable th) {
            Log.e(TAG, "configLmLib... ", th);
        } finally {
            CommonUtils.closeQuietly(is);
        }
    }

    public static boolean configLmLib(LocalMode lm, String config) {
        try {
            if (TextUtils.isEmpty(config)) {
                Logger.e(TAG, "Config for lib empty");
                return false;
            }
            ObjectMapper mapper = new ObjectMapper();
            JsonNode actualObj = mapper.readTree(config);
            boolean test = actualObj.has("content") && actualObj.get("content").has("config");

            boolean isConfigured = lm.configure(actualObj);

            Log.w(TAG, "LocalModeHelper test: " + test);
            Log.w(TAG, "configLmLib isConfigured: " + isConfigured + " conf size: " + config.length() + " data:" + config);
            return isConfigured;
        } catch (Throwable th) {
            Log.e(TAG, "configLmLib... ", th);
        }
        return false;
    }

    public static boolean initWithNewPathLibrary(File libFile) {
        try {
            Logger.e("### local lib initialization ###", "local lib init started with: " + libFile);
//      TODO: investigate how is working that delete procedure (lmlib) ???
            File path = ContextCompat.getCodeCacheDir(AppContext.get()).getAbsoluteFile();
            if (path.isDirectory()) {
                File[] files = path.listFiles();
//      TODO: do we need to check is files isNull
                for (File fileDel : files) {
                    boolean status = fileDel.delete();
                    Log.i(TAG, "fileDel:" + fileDel.getName() + " status:" + status);
                }
            }

            PrefUtils.setLibraryVersion(libFile);

            LocalMode lm = load(libFile);
            if (lm == null) return false;

            boolean status = false;
            String offlineConfigServer = ConfigPrefHelper.getPref().offlineConfigServer();
            if (!TextUtils.isEmpty(offlineConfigServer)) {
                status = configLmLib(lm, offlineConfigServer);
                ConfigPrefHelper.getPref().setOfflineConfig(offlineConfigServer);
            }

            MobileEntryApplication.getConfigPreferences().setLmlibFilePath(libFile.getAbsolutePath());
            Log.i(TAG, "initWithNewPathLibrary  status:" + status);
            sLocalMode = lm;
            return true;
        } catch (Throwable e) {
            Logger.w(TAG, "Failed to load lmlib", e);
        }
        return false;
    }

    private static LocalMode firstInitLibrary() {
        try {
            Context context = AppContext.get();
            File destFile = new File(context.getObbDir(), LM_LIB_NAME);
            CommonUtils.copyFileFromAssets(LM_LIB_NAME, destFile);
            PrefUtils.setLibraryVersion(destFile);
            LocalMode lm = load(destFile);
            MobileEntryApplication.getConfigPreferences().setLmlibFilePath(destFile.getPath()); // was AbsolutePath
            // offlineConfigServer
            if (TextUtils.isEmpty(ConfigPrefHelper.getPref().offlineConfig())) {
                initDefConfig();
            }
            return lm;
        } catch (Exception e) {
            Logger.w(TAG, "Failed to load lmlib", e);
        }
        return null;
    }

    private static LocalMode load() {
        String path = MobileEntryApplication.getConfigPreferences().lmlibFilePath();
        LocalMode lm;
        if (TextUtils.isEmpty(path)) {
            lm = firstInitLibrary();
        } else {
            lm = load(new File(path));
            PrefUtils.setLibraryVersion(path);
        }
        String offlineConfig = Optional.ofNullable(ConfigPrefHelper.getPref().offlineConfigServer())
                .filter(s -> !s.isEmpty())
                .orElse(ConfigPrefHelper.getPref().offlineConfig());
        configLmLib(lm, offlineConfig);
        return lm;
    }

    private static LocalMode load(File destFile) {
        try {
            Context context = AppContext.get();
            DexClassLoader classLoader = new DexClassLoader(destFile.getAbsolutePath(), ContextCompat.getCodeCacheDir(context).getAbsolutePath(), null, context.getClassLoader());
            Class<?> clazz = classLoader.loadClass(LOCAL_MODE_CLASS_NAME);
            return new LocalModeImpl(clazz, clazz.newInstance());
        } catch (Exception e) {
            Logger.w(TAG, "Failed to load lmlib", e);
        }
        return null;
    }
}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/model/LocalModeHelper.kt
package de.dimedis.mobileentry.model

import android.text.TextUtils
import android.util.Log
import androidx.core.content.ContextCompat
import com.fasterxml.jackson.databind.ObjectMapper
import dalvik.system.DexClassLoader
import de.dimedis.lmlib.LocalModeImpl
import de.dimedis.lmlib.myinterfaces.Access
import de.dimedis.lmlib.myinterfaces.LocalMode
import de.dimedis.mobileentry.ConfigPref
import de.dimedis.mobileentry.util.AppContext.get
import de.dimedis.mobileentry.util.CommonUtils
import de.dimedis.mobileentry.util.ConfigPrefHelper
import de.dimedis.mobileentry.util.Logger
import de.dimedis.mobileentry.util.Logger.w
import de.dimedis.mobileentry.util.PrefUtils
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.util.jar.JarFile

object LocalModeHelper {
    private val TAG = LocalModeHelper::class.java.simpleName
    private const val LOCAL_MODE_CLASS_NAME = "de.dimedis.lmlib.impl.LocalModeImpl"
    private const val LM_LIB_NAME = "lmlib.jar"
    private var sLocalMode: LocalMode? = null

    fun performEntry(ticketCode: String): Access? {
        return performEntry(PerfData().initFromPref(), ticketCode)
    }

    fun performEntry(pref: PerfData, ticketCode: String): Access? {
        val lm: LocalMode = getLocalMode()
        try {
            val access: Access? = lm.performEntry(ticketCode,
                pref.deviceSuid,
                pref.userSession,
                pref.userSuid,
                pref.userName,
                pref.fair,
                pref.border,
                pref.lang)
            Log.w(TAG, "access: $access")
            return access
        } catch (e: Exception) {
            w(TAG, "Failed to perform entry", e)
        }
        return null
    }

    fun performCheckout(ticketCode: String): Access? {
        return performCheckout(PerfData().initFromPref(), ticketCode)
    }

    fun performCheckout(pref: PerfData, ticketCode: String): Access? {
        val lm: LocalMode = getLocalMode()
        try {
            return lm.performCheckout(ticketCode,
                pref.deviceSuid,
                pref.userSession,
                pref.userSuid,
                pref.userName,
                pref.fair,
                pref.border,
                pref.lang)
        } catch (e: java.lang.Exception) {
            w(TAG, "Failed to perform checkout", e)
        }
        return null
    }

    fun recordEntry(ticketCode: String): Access? {
        val lm = getLocalMode()
        val border = ConfigPrefHelper.getUsersBorder()
        if (border != null) {
            try {
                return lm.recordEntry(ticketCode, ConfigPref.deviceSuid, ConfigPref.userSession, ConfigPref.userSuid, ConfigPref.userName, border.fair, border.border, ConfigPref.currentLanguage)
            } catch (e: Exception) {
                w(TAG, "Failed to record entry", e)
            }
        }
        return null
    }

    fun getLocalMode(): LocalMode {
        if (sLocalMode == null) {
            sLocalMode = load()
        }
        return sLocalMode as LocalMode
    }

    fun getLocalLibVersionFromManifestFile(fullPathToJar: File): String? {
        try {
            val jar = JarFile(fullPathToJar)
            val manifest = jar.manifest
            val mattr = manifest.mainAttributes
            val version = mattr.getValue("Implementation-Version")
            println("\\nVersion$version:\\n")
            if (version == null) {
                throw IOException("unable to read jar version from manifest")
            }
            return version
        } catch (fex: FileNotFoundException) {
            Log.e(TAG, "unable to find lib file: " + fullPathToJar.absolutePath)
        } catch (ex: IOException) {
            ex.printStackTrace()
        }
        return null
    }

    private fun initDefConfig() {
        var `is`: InputStream? = null
        try {
            `is` = get().assets.open("offline-test-config.json")
            val conf = CommonUtils.inputStreamToString(`is`)
            ConfigPref.offlineConfig = (conf)
            Log.w(TAG, "initDefConfig conf size: " + (conf?.length ?: "null") + " data:" + conf)
        } catch (th: Throwable) {
            Log.e(TAG, "configLmLib... ", th)
        } finally {
            CommonUtils.closeQuietly(`is`)
        }
    }

    fun configLmLib(lm: LocalMode, config: String): Boolean {
        try {
            if (TextUtils.isEmpty(config)) {
                Logger.e(TAG, "Config for lib empty")
                return false
            }
            val mapper = ObjectMapper()
            val actualObj = mapper.readTree(config)
            val test = actualObj.has("content") && actualObj["content"].has("config")
            val isConfigured = lm.configure(actualObj)
            Log.w(TAG, "LocalModeHelper test: $test")
            Log.w(TAG, "configLmLib isConfigured: " + isConfigured + " conf size: " + config.length + " data:" + config)
            return isConfigured
        } catch (th: Throwable) {
            Log.e(TAG, "configLmLib... ", th)
        }
        return false
    }

    fun initWithNewPathLibrary(libFile: File): Boolean {
        try {
            Logger.e("### local lib initialization ###", "local lib init started with: $libFile")

            val path = ContextCompat.getCodeCacheDir(get()).absoluteFile
            if (path.isDirectory) {
                val files = path.listFiles()
                for (fileDel in files!!) {
                    val status = fileDel.delete()
                    Log.i(TAG, "fileDel:" + fileDel.name + " status:" + status)
                }
            }
            PrefUtils.setLibraryVersion(libFile)
            val lm = load(libFile) ?: return false
            var status = false
            val offlineConfigServer = ConfigPref.offlineConfigServer
            if (!TextUtils.isEmpty(offlineConfigServer)) {
                status = offlineConfigServer?.let { configLmLib(lm, it) } == true
                ConfigPref.offlineConfig = (offlineConfigServer)
            }
            ConfigPref.lmlibFilePath = (libFile.absolutePath)
            Log.i(TAG, "initWithNewPathLibrary  status:$status")
            sLocalMode = lm
            return true
        } catch (e: Throwable) {
            w(TAG, "Failed to load lmlib", e)
        }
        return false
    }

    private fun firstInitLibrary(): LocalMode? {
        try {
            val context = get()
            val destFile = File(context.cacheDir, LM_LIB_NAME)
            CommonUtils.copyFileFromAssets(LM_LIB_NAME, destFile)
            PrefUtils.setLibraryVersion(destFile)
            val lm = load(destFile)
            ConfigPref.lmlibFilePath = (destFile.absolutePath)
            // offlineConfigServer
            if (TextUtils.isEmpty(ConfigPref.offlineConfig)) {
                initDefConfig()
            }
            return lm
        } catch (e: Exception) {
            w(TAG, "Failed to load lmlib", e)
        }
        return null
    }

    private fun load(): LocalMode {
        val path: String? = ConfigPref.lmlibFilePath
        val lm: LocalMode
        if (TextUtils.isEmpty(path)) {
            lm = firstInitLibrary()!!
        } else {
            lm = path?.let { File(it) }?.let { load(it) }!!
            PrefUtils.setLibraryVersion(path)
        }
        val offlineConfig = ConfigPref.offlineConfig
        if (offlineConfig != null) {
            configLmLib(lm, offlineConfig)
        }
        return lm
    }

    private fun load(destFile: File): LocalMode? {
        try {
            val context = get()
            val classLoader = DexClassLoader(
                destFile.absolutePath,
                ContextCompat.getCodeCacheDir(context).absolutePath,
                null,
                context.classLoader
            )
            val clazz = classLoader.loadClass(LOCAL_MODE_CLASS_NAME) as Class<*>
            return LocalModeImpl(clazz, clazz.newInstance())
        } catch (e: java.lang.Exception) {
            w(TAG, "Failed to load lmlib", e)
        }
        return null
    }
}